vaccine_feed_ingest/utils/match.py

```diff
@@ -1,8 +1,138 @@
 import re
 from typing import Optional
 
+import jellyfish
+import phonenumbers
+import us
 from vaccine_feed_ingest_schema import location
 
+from .log import getLogger
+from .normalize import provider_id_from_name
+
+logger = getLogger(__file__)
+
+
+def is_address_similar(
+    source: location.NormalizedLocation,
+    candidate: dict,
+) -> Optional[bool]:
+    """Compare address for similarity
+
+    - True if address matches exactly after canonicalization
+    - False if anything those match
+    - None if couldn't compare addresses
+    """
+    if not source.address:
+        return None
+
+    if not candidate["properties"].get("full_address"):
+        return None
+
+    # States must exist to compare
+    if not source.address.state or not candidate["properties"].get("state"):
+        return None
+
+    # City must exist to compare
+    if not source.address.city:
+        return None
+
+    # Street must exist to compare
+    if not source.address.street1:
+        return None
+
+    # States must exactly match
+    src_state = us.states.lookup(source.address.state)
+    cand_state = us.states.lookup(candidate["properties"]["state"])
+
+    if src_state != cand_state:
+        return False
+
+    src_addr = canonicalize_address(get_full_address(source.address))
+    cand_addr = canonicalize_address(candidate["properties"]["full_address"])
+
+    return src_addr == cand_addr
+
+
+def is_provider_similar(
+    source: location.NormalizedLocation,
+    candidate: dict,
+    threshold: float = 0.9,
+) -> Optional[float]:
+    """Calculate a similarity score between providers.
+
+    - True if similarity is above threshold
+    - False if similarity is below threshold
+    - None if couldn't compare providers
+    """
+    if not source.parent_organization:
+        return None
+
+    src_org = source.parent_organization.name or source.parent_organization.id
```

Comment:
Is this needed because normalizers use these fields inconsistently?

Reply:
Yes. Some of the normalizers just fill out the name.

This is also a little hacky, I would like to match directly to providers, but VIAL doesn't return a strong identifier for their providers in the API yet.